lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1243 @@
+		//   - MUST accept P2WSH, P2WPKH, P2TR scripts
+		//   - MAY fail the negotiation if script is non-standard
+		if !msg.script.is_v0_p2wpkh() && !msg.script.is_v0_p2wsh() && !msg.script.is_v1_p2tr() {
+			return Err(AbortReason::InvalidOutputScript);
+		}
+
+		if self.outputs.iter().any(|(serial_id, _)| *serial_id == msg.serial_id) {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the `serial_id` is already included in the transaction
+			return Err(AbortReason::DuplicateSerialId);
+		}
+
+		let output = TxOut { value: msg.sats, script_pubkey: msg.script.clone() };
+		self.outputs.entry(msg.serial_id).or_insert(output);
+		Ok(())
+	}
+
+	fn remote_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) -> Result<(), AbortReason> {
+		if !self.is_serial_id_valid_for_counterparty(&msg.serial_id) {
+			return Err(AbortReason::IncorrectSerialIdParity);
+		}
+		if let Some(_) = self.outputs.remove(&msg.serial_id) {
+			Ok(())
+		} else {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the input or output identified by the `serial_id` was not added by the sender
+			//    - the `serial_id` does not correspond to a currently added input
+			Err(AbortReason::SerialIdUnknown)
+		}
+	}
+
+	fn local_tx_add_input(&mut self, msg: &msgs::TxAddInput) {
+		let tx = msg.prevtx.as_transaction();
+		let input = TxIn {
+			previous_output: OutPoint { txid: tx.txid(), vout: msg.prevtx_out },
+			sequence: Sequence(msg.sequence),
+			..Default::default()
+		};
+		debug_assert!((msg.prevtx_out as usize) < tx.output.len());
+		let prev_output = &tx.output[msg.prevtx_out as usize];
+		self.prevtx_outpoints.insert(input.previous_output.clone());
+		self.inputs.insert(
+			msg.serial_id,
+			TxInputWithPrevOutput { input, prev_output: prev_output.clone() },
+		);
+	}
+
+	fn local_tx_add_output(&mut self, msg: &msgs::TxAddOutput) {
+		self.outputs
+			.insert(msg.serial_id, TxOut { value: msg.sats, script_pubkey: msg.script.clone() });
+	}
+
+	fn local_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) {
+		self.inputs.remove(&msg.serial_id);
+	}
+
+	fn local_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) {
+		self.outputs.remove(&msg.serial_id);
+	}
+
+	fn build_transaction(mut self) -> Result<Transaction, AbortReason> {
+		// The receiving node:
+		// MUST fail the negotiation if:
+
+		// - the peer's total input satoshis is less than their outputs
+		let counterparty_inputs_value: u64 =
+			self.counterparty_inputs_contributed().map(|input| input.prev_output.value).sum();
+		let counterparty_outputs_value =
+			self.counterparty_outputs_contributed().map(|output| output.value).sum();
+		if counterparty_inputs_value < counterparty_outputs_value {
+			return Err(AbortReason::OutputsValueExceedsInputsValue);
+		}
+
+		// - there are more than 252 inputs
+		// - there are more than 252 outputs
+		if self.inputs.len() > MAX_INPUTS_OUTPUTS_COUNT
+			|| self.outputs.len() > MAX_INPUTS_OUTPUTS_COUNT
```

Comment:
I don't think we can at that point since the counterparty could still decide to remove some and later be under the maximum at the end of the negotiation since the could still send a bunch more messages until they reach the message count limit.